klippy/extras/safe_z_home.py

```diff
@@ -0,0 +1,85 @@
+# Perform Z Homing at specific XY coordinates.
+#
+# Copyright (C) 2019 Florian Heilmann <Florian.Heilmann@gmx.net>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+import probe, toolhead
+import logging
+
+class SafeZHoming:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        try:
+            (self.home_x_pos,
+            self.home_y_pos) = config.get("home_xy_position",
+                                            default=",").split(',')
+        except:
+            raise config.error("Unable to parse home_xy_position in %s" % (
+                config.get_name()))
+
+        self.z_hop = config.getfloat("z_hop", default=0.0)
+        self.z_hop_speed = config.getfloat('z_hop_speed', 15., above=0.)
+        self.speed = config.getfloat('speed', 50.0, above=0.)
+        self.gcode = self.printer.lookup_object('gcode')
+        self.toolhead = None
+        self.kinematics = None
+        self.gcode.register_command("G28", None)
+        self.gcode.register_command("G28", self.cmd_G28)
+
+        if self.printer.lookup_object("homing_override", default=None):
+            raise config.error("homing_override and safe_z_homing cannot"
+                                    +" be used simultaneously")
+
+    def cmd_G28(self, params):
+        if not self.toolhead:
+            self.toolhead = self.printer.lookup_object('toolhead')
+            self.kinematics = self.toolhead.get_kinematics()
+
+        # Perform Z Hop if necessary
+        if self.z_hop != 0.0:
+            # Check if the zhop would exceed the printer limits
+            pos = self.toolhead.get_position()
+            kin_status = self.kinematics.get_status()
+            if (kin_status['homed'][-1] and
+                pos[2] + self.z_hop > kin_status['limits'][-1][1]):
```

Comment:
I'm not sure it's a good idea to expose the kinematic "limits" as that's pretty low-level.  I think you should be able to get at the maximum Z position during init with something like: `self.max_z = config.getsection('stepper_z').getfloat('position_max')`

If the z axis is homed, I think this code still needs to check if the z position is below z_hop (and if so, to move up to a minimum of z=z_hop).

Reply:
While this works on cartesian and corexy, it won't work on, e.g. deltas. Do you have a suggestion how I can unify all the kinematics in a clean way? 

One possibility could be to, for now, limit ourselves to the Z axis, as that information is pretty easy to extract from each kinematic.

Reply:
Also the check will ignore the zhop if the following two conditions are met:

z is homed 
current_z + zhop > limit_z

In all other cases the hop is executed. I'm having a hard time imagining where a zhop would exceed the limit without current_z being at least Z-hop (except for super-flat machines or machines with an outrageous zhop). If you'd still like me to put in this change, let me know and I'll fix it.

Reply:
> While this works on cartesian and corexy, it won't work on, e.g. deltas. Do you have a suggestion how I can unify all the kinematics in a clean way?

I'm not sure it makes sense to perform "safe z homing" on a printer without a z axis.  So, I think it would be fine to raise a config error if there is no stepper_z config section.

> In all other cases the hop is executed.

Sorry - I misread the original code.  What you have now looks fine to me.

-Kevin